crates/iceberg/src/scan.rs

```diff
@@ -155,8 +189,22 @@ impl TableScan {
             .await?;
 
             // Generate data file stream
-            let mut entries = iter(manifest_list.entries());
-            while let Some(entry) = entries.next().await {
+            for entry in manifest_list.entries() {
+                // If this scan has a filter, check the partition evaluator cache for an existing
+                // PartitionEvaluator that matches this manifest's partition spec ID.
+                // Use one from the cache if there is one. If not, create one, put it in
+                // the cache, and take a reference to it.
+                if let Some(filter) = self.filter.as_ref() {
+                    let partition_evaluator = partition_evaluator_cache
+                            .entry(entry.partition_spec_id())
+                            .or_insert_with_key(|key| self.create_partition_evaluator(key, filter));
+
+                    // reject any manifest files whose partition values don't match the filter.
+                    if !partition_evaluator.filter_manifest_file(entry) {
```

Comment:
I think we should apply `ManifestEvaluator` independently from `PartitionEvaluator` (see comment down below in struct PartitionEvaluator).

Also, I'm not sure if we can simply 'skip' the file with continue. In the python impl the Partition- and MetricsEvaluator are applied on each DataFile on multiple threads - so I believe we have to collect all relevant manifest files before and then apply the Partition- and MetricsEvaluator on the complete collections (or split the work into multiple threads)?